Refactor code structure of relay contract and support storage migration.

riversyang opened the pull request:
1. Refactor code structure of OctopusRelay contract. Mainly separate logic related to single app chain to module 'appchain'.
2. Add trait 'AppchainPipeline' in module 'pipeline' to gather functions related to app chain pipeline actions into module 'pipeline'.
3. Add module 'bridging' for holding functions related to cross-chain token lock and unlock (keeping existed logic).
4. Use 'LazyOption' to wrap struct values in collections for supporting contract storage migration.
5. Add module 'storage_key' for maintaining storage keys using in contract, which will be useful in case of changing fields of struct is necessary.
6. Add sample implementation (module 'storage_migration') and tests of contract storage migration for further reference. 
7. Add burnt gas log to tests.
8. Other minor changes related to the above changes.
9. Upgrade near-sdk to version 3.1.0.

riversyang committed:
Refactor relay contract into modules.

riversyang committed:
Fix bug of function 'get_current_validator_set'.

riversyang committed:
Merge branch 'master' into refactor-code-structure

riversyang committed:
Update .gitignore

riversyang committed:
Fix bug of without saving collection element when its content changed.

riversyang committed:
Update state.rs

riversyang committed:
Update lib.rs

riversyang committed:
Fix bug for passing all tests.

riversyang committed:
Optimize implementation for storage migration.

riversyang committed:
Complete refactoring.

1. Remove the field 'note' which is for testing, from struct 'AppchainValidator'.
2. Comment out module 'storage_migration' of crate 'octopus-relay'.
3. Comment out testing function for storage migration.

riversyang committed:
Upgrade near-sdk to version 3.1.0.

en commented on octopus-relay/src/appchain/metadata.rs:
```diff
@@ -0,0 +1,124 @@
+#[serde(crate = "near_sdk::serde")]
+pub struct AppchainMetadata {
+    /// Appchain id
+    pub id: AppchainId,
```
There's no need to use a type alias here, the appchain_id must be a `String`.

riversyang replied:
This is to follow the original style of relay contract before refactoring. We can unify the style in whole project later.

madwiki replied:
Use alias to make it clearer during development, for example:

pub fn get_validator(
        appchain_id: String,
        validator_id: String,
)
which is the validator_id string? near account or appchain account(hex string)?
So we use AppchainId, ValidatorId, AccountId to distinguish.

en commented on octopus-relay/src/appchain/state.rs:
```diff
@@ -0,0 +1,397 @@
+    /// Id of the appchain
+    pub id: AppchainId,
+    /// Validators collection of the appchain
+    pub validators: UnorderedMap<ValidatorId, LazyOption<AppchainValidator>>,
```
I have a question about this type.
Validators of appchain need to agree on the validator_set they observed.
But if some of the validators observe a set `{a, b, c}`, and others observe `{a, c, b}`, there is no way to reach a consensus.
Although we can sort the set on the appchain side, it's better if the set is ordered.

riversyang replied:
Got it.
Considering we can also remove validator from validator set, we can NOT ensure the order of validators in storage on chain even if we use an extra Vector to store their ids (because the mechanism of removing element from array is to move the last element to the position of removed one). 
So,  I'll add sorting logic when return validator set out of contract.

en commented on octopus-relay/src/appchain/state.rs:
```diff
@@ -0,0 +1,397 @@
+    /// Total staked balance of OCT token of the appchain
+    pub staked_balance: Balance,
+    /// Total upvote balance of OCT token of the appchain
+    pub upvote_balace: Balance,
```
typo

riversyang replied:
Fixed.

riversyang committed:
Update state.rs

Rename 'id' of AppchainState to 'appchain_id'.
Fix typo of field 'upvote_balance'.

madwiki committed:
fix bug for stake_more

riversyang committed:
Correct logic related to facts of appchain.

riversyang committed:
Sort validators by their ids before return validator set out of contract.
